### ISSUE TITLE
OV9281 question

### ISSUE THREAD

**Issue by ruddy17** [Banner13/ov9281#1]
Hi, Banner13. Sorry to try reaching you through issues, consider it a DM :) I am developing a device with an OV9281 and lack of proper documentation is killing me. Could you please explain in short, what mode do registers in your driver enable? I seek an example for DVP output, maybe your code could serve me as an example.

**Comment by Banner13** [comment]
OK, I have a reference manual on ov9281 in hand, but it is not very complete either, I can provide it to you later. About the questions you have in development, I will try to answer them, but the results may fly over the earth. thank you for having followed me

**Comment by ruddy17** [comment]
Yeah, I got the "datasheet: product specification" which is probably the same you have (139 pages?).  Not even closely thorough enough.

I am confused with proper PLL configuration. I need to receive on STM32, which can't keep up with default 100mhz pix clk. I tried to get working PLL settings, and thus far I found out that in order to lower pixclk I have to lower all the other clocks proportionally, otherwise I get missing pixels and lines (as if the resolution is wrong, like 511 lines instead of 800). 

Somehow I managed to get seemingly proper frame (looking with logic analyzer) only at 640x400, with all pixels and lines intact. But when actually receiving and decoding the captured frame, I get strange repeating artifacts in the middle of the frame, like some broken pattern in front of an actual image from sensor. I suspect that it may still be due to some inconsistency between different clocks. 

By the way, I made a PLL calculator, can share if you're interested.

Do you know by any chance, is there a better method than blind trial and error for picking the right PLL settings? Also, do you have to change anything else after tinkering with PLLs, like frame blanks etc.?

**Comment by Banner13** [comment]
Emmm , how do you set registers about PLL?



At 2022-09-20 15:45:29, "Sergey Omelchenko" ***@***.***> wrote:

Yeah, I got the "datasheet: product specification" which is probably the same you have (139 pages?). Not even closely thorough enough.

I am confused with proper PLL configuration. I need to receive on STM32, which can't keep up with default 100mhz pix clk. I tried to get working PLL settings, and thus far I found out that in order to lower pixclk I have to lower all the other clocks proportionally, otherwise I get missing pixels and lines (as if the resolution is wrong, like 511 lines instead of 800).

Somehow I managed to get seemingly proper frame (looking with logic analyzer) only at 640x400, with all pixels and lines intact. But when actually receiving and decoding the captured frame, I get strange repeating artifacts in the middle of the frame, like some broken pattern in front of an actual image from sensor. I suspect that it may still be due to some inconsistency between different clocks.

By the way, I made a PLL calculator, can share if you're interested.

Do you know by any chance, is there a better method than blind trial and error for picking the right PLL settings? Also, do you have to change anything else after tinkering with PLLs, like frame blanks etc.?

**Comment by ruddy17** [comment]
![изображение](https://user-images.githubusercontent.com/4340532/191440731-c3fc9679-54ae-4dd3-8f77-e5c88410abdd.png)
Using my calculator (self written) which generates register values from selected dividers. Then I just add pll-related registers to initialization array. Communication with the sensor is working ok, I thoroughly tested that of course.

If I only reduce pixclk, then framing gets broken (wrong number of pixels and lines goes from sensor). If I reduce all frequencies proportionally, then framing seems ok (only at 640x400, still broken on full frame). But even then some pixels get glitchy.

**Comment by Banner13** [comment]
At 2022-09-12 16:34:44, "Sergey Omelchenko" ***@***.***> wrote:

Hi, Banner13. Sorry to try reaching you through issues, consider it a DM :) I am developing a device with an OV9281 and lack of proper documentation is killing me. Could you please explain in short, what mode do registers in your driver enable? I seek an example for DVP output, maybe your code could serve me as an example.
